tests/integration.rs

```diff
@@ -743,3 +745,65 @@ fn test_expire() -> Result<()> {
 
     Ok(())
 }
+
+#[test]
+fn test_defrag() -> Result<()> {
+    let port: u16 = 6503;
+    let _guards = vec![start_redis_server_with_module("data_type", port)
+        .with_context(|| "failed to start redis server")?];
+    let mut con =
+        get_redis_connection(port).with_context(|| "failed to connect to redis server")?;
+
+    // Configure active defrag
+    redis::cmd("config")
+        .arg(&["set", "hz", "100"])
+        .query(&mut con)
+        .with_context(|| "failed to run 'config set hz 100'")?;
+
+    redis::cmd("config")
+        .arg(&["set", "active-defrag-ignore-bytes", "1"])
+        .query(&mut con)
+        .with_context(|| "failed to run 'config set active-defrag-ignore-bytes 1'")?;
+
+    redis::cmd("config")
+        .arg(&["set", "active-defrag-threshold-lower", "0"])
+        .query(&mut con)
+        .with_context(|| "failed to run 'config set active-defrag-threshold-lower 0'")?;
+
+    redis::cmd("config")
+        .arg(&["set", "active-defrag-cycle-min", "99"])
+        .query(&mut con)
+        .with_context(|| "failed to run 'config set active-defrag-cycle-min 99'")?;
+
+    // enable active defrag
+    if redis::cmd("config")
+        .arg(&["set", "activedefrag", "yes"])
+        .query::<String>(&mut con)
+        .is_err()
+    {
+        // Server the does not support active defrag, avoid failing the test.
+        return Ok(());
+    }
+
+    let start = SystemTime::now();
+    loop {
+        let res: HashMap<String, usize> = redis::cmd("alloc.defragstats")
+            .query(&mut con)
+            .with_context(|| "failed to run 'config set active-defrag-cycle-min 99'")?;
+        let num_defrag_globals = res.get("num_defrag_globals").ok_or_else(|| {
+            anyhow::Error::msg("Failed getting 'num_defrag_globals' value from result")
+        })?;
+        // Wait till we will get at least 2 defrag cycles.
+        // We are looking at num_defrag_globals because this is supported by all Redis versions
+        // that supports defrag.
+        if *num_defrag_globals > 2 {
+            break;
+        }
+        let duration = SystemTime::now().duration_since(start)?;
+        if duration > Duration::from_secs(30) {
+            return Err(anyhow::Error::msg("Failed waiting for defrag cycle"));
+        }
+    }
```

Comment:
Just to clarify the test case. We are testing here that the process works fine and that we have the two defrag cycles. Is it possible for the tests to ever reach (in practice of in theory) the code which checks for the timeout of `30 seconds`? I wonder when and in what cases this might happen. How can we reduce the possible flakiness of this test?

Reply:
> Is it possible for the tests to ever reach (in practice of in theory) the code which checks for the timeout of 30 seconds?

Yes, if we have a bug (or a bug in Redis), the defrag functions will not be called and we will reach a timeout.

> How can we reduce the possible flakiness of this test?

I configured the degrag to be very intensive so 30 seconds is much more than what is actually needed and if we reach it we must have a bug.

Reply:
Thanks.